draft-ietf-quic-tls.md

```diff
@@ -1101,75 +1101,153 @@ anticipation of receiving a ClientHello.
 If an endpoint detects a second update before it has sent any packets with
 updated keys, it indicates that its peer has updated keys twice without awaiting
-a reciprocal update.  An endpoint MUST treat consecutive key updates as a fatal
-error and abort the connection.
+a reciprocal update.  An endpoint MUST treat consecutive key updates as a
+connection error of type KEY_UDPATE_ERROR.
 
-An endpoint SHOULD retain old keys for a period of no more than three times the
-Probe Timeout (PTO, see {{QUIC-RECOVERY}}).  After this period, old keys and
-their corresponding secrets SHOULD be discarded.  Retaining keys allow endpoints
-to process packets that were sent with old keys and delayed in the network.
-Packets with higher packet numbers always use the updated keys and MUST NOT be
-decrypted with old keys.
 
-This ensures that once the handshake is complete, packets with the same
-KEY_PHASE will have the same packet protection keys, unless there are multiple
-key updates in a short time frame succession and significant packet reordering.
+## Using Old Keys {#key-update-old-keys}
 
-~~~
-   Initiating Peer                    Responding Peer
+If the most recent packet sent by the endpoint contained a Key Update Permitted
+bit set to 0, a Key Phase other than the value expected indicates that the
+packet was protected with old keys.  If those old keys are available, then they
+can be used to remove packet protection.
 
-@M QUIC Frames
-               New Keys -> @N
-@N QUIC Frames
-                      -------->
-                                          QUIC Frames @M
-                          New Keys -> @N
-                                          QUIC Frames @N
-                      <--------
-~~~
-{: #ex-key-update title="Key Update"}
+Old keys might be used to remove protection from packets that were are reordered
+in the network.  However, it is never valid for old keys to be used to protect
+packets with packets that have higher packet numbers than packets that were
+protected with newer keys.  An endpoint that successfully removes protection
+with old keys when newer keys were used for packets with lower packet numbers
+MUST treat this as a connection error of type KEY_UPDATE_ERROR.
 
-A packet that triggers a key update could arrive after successfully processing a
-packet with a higher packet number.  This is only possible if there is a key
-compromise and an attack, or if the peer is incorrectly reverting to use of old
-keys.  Because the latter cannot be differentiated from an attack, an endpoint
-MUST immediately terminate the connection if it detects this condition.
+An endpoint SHOULD retain old read keys for a period of no more than three times
+the Probe Timeout (PTO, see {{QUIC-RECOVERY}}).  After this period, old read
+keys and their corresponding secrets SHOULD be discarded.  An endpoint MAY keep
+the Key Update Permitted bit set to 0 until it discards old read keys to limit
```

Comment:
How about explicitly stating "until it discards the 0-RTT key or the previous 1-RTT read key" here, to be clear that we can have the Key Update Permitted bit set to zero until the 0-RTT key is discarded when 0-RTT is used.